### PR TITLE
Improve support for TYPO3 schema management

### PR DESCRIPTION
**Pull request by DanielSiepmann**
Support multiple databases.
Fetch all relevant schemas from all databases.
Pass them back to TYPO3.
TYPO3 will check the mapping configuration and keep the schemas as they are.

We also remove the constraint part from foreign keys, as that's the only part TYPO3 can't work with.

Resolves: #197

### CODE REVIEW

**Comment by DanielSiepmann** [comment]
There you go. I've added your feedback. I couldn't check the code base, as our customer still uses 2021.10 instead of current master. Looking forward to read your feedback on the proposed changes.

File: Classes/Setup.php [hunk 1]
```diff
@@ -163,28 +163,38 @@ public static function execute()
...
-        $conn = self::context()->db();
+        $ctx = self::getContext();
+        $dbm = $ctx->getDatabaseManager();
+        $connectionNames = array_keys($ctx->getConfig()->get( 'resource'));
+        $connectionNames = array_filter($connectionNames, fn (string $key): bool => str_starts_with($key, 'db'));
 
-        foreach (['fe_users_', 'madmin_', 'mshop_'] as $prefix) {
-            $result = $conn->create('SHOW TABLES like \'' . $prefix . '%\'')->execute();
+        foreach ($connectionNames as $connectionName) {
+            $conn = $dbm->acquire($connectionName);
```
Comment (aimeos) [inline Classes/Setup.php:172]: It's just `$conn = $ctx->db($connectionName);` in the current version

File: Classes/Setup.php [hunk 2]
```diff
@@ -163,28 +163,38 @@ public static function execute()
...
-        foreach ($tables as $table) {
-            $result = $conn->create('SHOW CREATE TABLE ' . $table)->execute();
+            foreach($tables as $table) {
+                $result = $conn->create('SHOW CREATE TABLE `' . $table . '`')->execute();
 
-            while (($row = $result->fetch(\Aimeos\Base\DB\Result\Base::FETCH_NUM)) !== null) {
-                $str = preg_replace('/,[\n ]*CONSTRAINT.+CASCADE/', '', $row[1]);
-                $str = preg_replace('/ DEFAULT CHARSET=[^ ;]+/', '', $str);
-                $str = preg_replace('/ COLLATE=[^ ;]+/', '', $str);
-                $str = str_replace('"', '`', $str);
```
Comment (aimeos) [inline Classes/Setup.php:184]: You need to keep that lines

Comment (DanielSiepmann) [reply Classes/Setup.php:184]: The adjustment worked fine for us in v11.5 LTS. What's the reason to keep the lines?

File: Classes/Setup.php (hunk 1 above)
Comment (DanielSiepmann) [reply Classes/Setup.php]: I also removed fetching the database manager and releasing the connection. It looked like the API to retrieve the manager no longer existed.

File: Classes/Setup.php (hunk 2 above)
Comment (aimeos) [reply Classes/Setup.php:184]: There had been installations which failed to upgrade when the default charset and collation was part of the output.

File: Classes/Setup.php (hunk 1 above)
Comment (aimeos) [reply Classes/Setup.php]: Yes, API changed and is much simpler now :-)

File: Classes/Setup.php (hunk 2 above)
Comment (DanielSiepmann) [reply Classes/Setup.php:184]: I'll adjust and test in our setup.

Comment (DanielSiepmann) [reply Classes/Setup.php:184]: TYPO3 v11.5 will drop the foreign key when revmoing the whole key:
```sql
 ALTER TABLE `mshop_supplier_list` DROP FOREIGN KEY `fk_mssupli_pid`
 ```

I therefore kept the removal of charset and collatin, but reduced the removal of the constraint/cascade to the necessary portion.
This extension is marked to be compatible with 11.5 only, so this should be fine I guess.

Comment (aimeos) [reply Classes/Setup.php:184]: Is that really related? Fear, that `$str = preg_replace('/CONSTRAINT `[a-zA-Z0-9_-]+` /', '', $str);` will only remove the first part of the constraint and `ON UPDATE CASCADE ON DELETE CASCADE` will be left over and then, TYPO3 will throw an error ...

Comment (DanielSiepmann) [reply Classes/Setup.php:184]: I've executed the code on every single change. TYPO3 v11.5 can properly parse foreign keys, it just can't handle the constraint itself.

Comment (aimeos) [reply Classes/Setup.php:184]: OK, let's try an see ...

**Comment by aimeos** [comment]
Thanks for your PR!

File: Classes/Setup.php (hunk 2 above)
Comment (DanielSiepmann) [reply Classes/Setup.php:184]: You can add tests covering the scenario which you then can execute against different dbms and TYPO3 versions.

Comment (aimeos) [reply Classes/Setup.php:184]: Help for updating the tests is very welcome! :-)

**Pull request merged**